CollectionAsserts.cs

```diff
@@ -135,6 +142,14 @@ public static void Contains<T>(IEnumerable<T> collection, Predicate<T> filter)
         /// <exception cref="DoesNotContainException">Thrown when the object is present inside the container</exception>
         public static void DoesNotContain<T>(T expected, IEnumerable<T> collection)
         {
+            // If an equality comparer is not explicitly provided, call into ICollection<T>.Contains which may
+            // use the collection's equality comparer for types like HashSet and Dictionary.
+            var icollection = collection as ICollection<T>;
+            if (icollection != null && icollection.Contains(expected))
+                throw new DoesNotContainException(expected, collection);
+
+            // We don't throw only if both ICollection<T>.Contains and our custom equality comparer say the collection
```

Comment:
`Contains` doesn't throw if either of `IC<T>.Contains` or the custom comparer say yes. The condition here has to be the logical negation of that so that it's not possible to have an item for which `Assert.Contains(item, c)` and `Assert.DoesNotContain(item, c)` both succeed.